pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,292 @@
+
+		// If the secret does not exist, create it
+		_, err = c.secretsClient.Create(ctx, &ngrok.SecretCreate{
+			VaultID:     vault.ID,
+			Name:        data.GetRemoteKey(),
+			Value:       string(value),
+			Metadata:    string(metadataJSON),
+			Description: psmd.Description,
+		})
+		return err
+	}
+
+	// If the secret exists, update it
+	_, err = c.secretsClient.Update(ctx, &ngrok.SecretUpdate{
+		ID:          existingSecret.ID,
+		Value:       ptr.To(string(value)),
+		Metadata:    ptr.To(string(metadataJSON)),
+		Description: ptr.To(psmd.Description),
+	})
+	return err
+}
+
+func (c *client) SecretExists(ctx context.Context, ref esv1.PushSecretRemoteRef) (bool, error) {
+	// Implementation for checking if a secret exists in ngrok
+	secret, err := c.getSecretByVaultNameAndSecretName(ctx, c.vaultName, ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, fmt.Errorf("error fetching secret: %w", err)
+	}
+
+	return (secret != nil), nil
+}
+
+// DeleteSecret deletes a secret from ngrok by its reference.
+func (c *client) DeleteSecret(ctx context.Context, ref esv1.PushSecretRemoteRef) error {
+	secret, err := c.getSecretByVaultNameAndSecretName(ctx, c.vaultName, ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		// If the secret or vault do not exist, we can consider it deleted.
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if secret == nil {
+		return nil
+	}
+
+	return c.secretsClient.Delete(ctx, secret.ID)
+}
+
+func (c *client) Validate() (esv1.ValidationResult, error) {
+	// Validate the client can list secrets with a timeout. If we
+	// can list secrets, we assume the client is valid(API keys, URL, etc.)
+	iter := c.secretsClient.List(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for iter.Next(ctx) {
+		return esv1.ValidationResultReady, nil
+	}
+
+	if iter.Err() != nil {
+		return esv1.ValidationResultError, fmt.Errorf("store is not allowed to list secrets: %w", iter.Err())
+	}
+
+	return esv1.ValidationResultReady, nil
+}
+
+func (c *client) GetSecret(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	// Implementation for getting a secret from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetSecretMap(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	// Implementation for getting a map of secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetAllSecrets(ctx context.Context, _ esv1.ExternalSecretFind) (map[string][]byte, error) {
+	// Implementation for getting all secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) Close(_ context.Context) error {
+	return nil
+}
+
+func (c *client) getVaultByName(ctx context.Context, name string) (*ngrok.Vault, error) {
```

Comment:
This needs to be cached somehow. On each call we do a vault by name call which lists all vault indifferently. This is not very efficient especially not if we only care about one vault. So do a vault call in the begin when the client is created then store it and use that instead. There is a Get call with id for vaults as well, might implement/use that for the vault client.

Reply:
The ngrok API doesn't currently support retrieval by name and does require listing all vaults and all secrets in order to find one by name as you mentioned. I'll work on implementing a cache client to be more efficient.

Reply:
I don't it needs to be that complex? Just save the result or the id and then do just get calls instead of list calls? I didn't mean to implement a complete LRU.

Reply:
This should also be fixed. One edge case I can think of is that ngrok vaults can be renamed. Since our API still doesn't support getting by name, I added a check before these to try to fetch the vault by ID to validate that the name still matches before syncing the secret. In the happy path, this adds an extra GET call before syncing a secret. In the unhappy path, this requires listing all of the vaults to find the new vault ID. IMO, this matches the users intent of the SecretStore/ClusterSecretStore with a vaultName.

If you think this should be handled differently, please let me know.